pxtwinrt/serial.ts

```diff
@@ -1,51 +1,84 @@
 
-        const filter = new RegExp(pxt.appTarget.serial.nameFilter);
-        const serialDeviceSelector = (Windows.Devices as any).SerialCommunication.SerialDevice.getDeviceSelector();
-        // Create a device watcher to look for instances of the Serial device
-        // The createWatcher() takes a string only when you provide it two arguments, so be sure to include an array as a second
-        // parameter (JavaScript can only recognize overloaded functions with different numbers of parameters).
-        watcher = Windows.Devices.Enumeration.DeviceInformation.createWatcher(serialDeviceSelector, [] as any);
-        watcher.addEventListener("added", (dis: any) => {
-            toArray(dis.detail).forEach((di: any) => {
-                if (!filter.test(di.name)) return;
-                pxt.debug(`serial port added ${di.name} - ${di.id}`);
-                ports[di.id] = {
-                    info: di
-                };
-                (Windows.Devices as any).SerialCommunication.SerialDevice.fromIdAsync(di.id)
-                    .done((dev: any) => {
-                        ports[di.id].device = dev;
-                        startDevice(di.id);
-                    })
-            });
-        });
-        watcher.addEventListener("removed", (dis: any) => {
-            toArray(dis.detail).forEach((di: any) => delete ports[di.id]);
-        })
-        watcher.addEventListener("updated", (dis: any) => {
-            toArray(dis.detail).forEach((di: any) => ports[di.id] ? ports[di.id].info.update(di.info) : null);
-        })
+        if (!watcher) {
+            deviceNameFilter = new RegExp(pxt.appTarget.serial.nameFilter);
+            const serialDeviceSelector = Windows.Devices.SerialCommunication.SerialDevice.getDeviceSelector();
+            // Create a device watcher to look for instances of the Serial device
+            // As per MSDN doc, to use the correct overload, we pass null as 2nd argument
+            watcher = Windows.Devices.Enumeration.DeviceInformation.createWatcher(serialDeviceSelector, null);
+        }
+
+        watcher.addEventListener("added", deviceAdded);
+        watcher.addEventListener("removed", deviceRemoved);
+        watcher.addEventListener("updated", deviceUpdated);
         watcher.start();
     }
 
+    export function suspendSerial() {
+        if (watcher) {
+            watcher.stop();
+            watcher.removeEventListener("added", deviceAdded);
+            watcher.removeEventListener("removed", deviceRemoved);
+            watcher.removeEventListener("updated", deviceUpdated);
+        }
+
+        Object.keys(activePorts).forEach((deviceId) => {
+            const port = activePorts[deviceId];
+            const device = port.device;
+            delete activePorts[deviceId];
+            device.close();
+        });
+        activePorts = {};
+    }
+
+    function deviceAdded(deviceInfo: DeviceInfo) {
+        if (!deviceNameFilter || !deviceNameFilter.test(deviceInfo.name)) {
```

Comment:
we should use the vendor id instead of the name hack

Reply:
Those properties are not default on the `DeviceInformation` object, so it's hard to do it in master. For now I'll leave the name filter and when I port these to V0 I'll debug with micro:bits attached, which will let me debug to see if I have access to that info